docs/adrs/0003-bulk-import-ui.md

```diff
@@ -0,0 +1,98 @@
+
+## Author
+
+@briehl
+
+## Status
+
+In progress
+
+## Background
+
+App Cells (and other cells) are all built using the [Jupyter front end extension framework](https://jupyter-notebook.readthedocs.io/en/stable/extending/frontend_extensions.html). On application and page load, this framework allows the overriding and customization of most FE components. Specifically, in the Narrative, this framework is only used for Cells. 
+
+Jupyter Notebook (and thus Narrative) cells come in three types - Code, Markdown, and Raw. All KBase cell extensions operate on Jupyter Code cells. It's easiest to think of them as subclassing - an App Cell is a Code Cell specialized for running Narrative Apps. Each of these extensions is activated by creating a cell with a specific metadata component that describes how the cell should be treated. For example, adding the following block to code cell metadata tells the extension machinery to treat it as a KBase App Cell:
+```
+{
+    kbase: {
+        type: 'app'
+    }
+}
+```
+
+The App Cell is, by far, the most complicated extension cell type. It has several components that all serve different purposes:
+* A finite state machine (FSM) that's used to determine cell state.
+* A core system used to render the cell based on state.
+* A message bus that's used to communicate with the rest of the FE application, as well as the BE (kernel) to get job state.
+* An interpreter / manager for app specs - turns the app specs into sets of input form elements.
+* A code translator - turns the various sets of inputs into code that can be executed to run the App.
+
+The reason for the App cell being an extension to code cells is the last point above: in essence, the primary UI components of the App cell serve as a fancy code generator. User interactions with form elements generate the code that can be used to start apps and view their results. Thus, anything that an app cell does with the UI can be done programmatically by an advanced user.
+
+> It should also be noted here that, regardless of the choice made, we may need to alter how app specs are made, interpreted, and rendered. The bulk import design has various inputs and outputs listed in horizontal rows while the existing app specs are rendered vertically. There's no notion of rows/columns in the app spec right now or how to visually display various inputs along a grid pattern. In order to effectively match the design we may need to do one of the following:
+> 1. Alter the app specs to support multiple in-line parameters while retaining backward compatibility (including documentation, etc.).
+> 2. Use a distinct app spec for importers that can include these options.
+> 3. Hard-code how all import apps look and feel in the Narrative interface itself.
+>
+> If we decide to update how app specs work, that means working in the `narrative_method_store` and `catalog` repos as well as internal documentation.
```

Comment:
Does it really require catalog changes? I don't remember making catalog changes the last time I added fields to the spec, but that was a while ago.

It'd be great to make it easier to update NMS to add fields to the spec, but that's probably out of scope.

Reply:
> Does it really require catalog changes? I don't remember making catalog changes the last time I added fields to the spec, but that was a while ago.

For changing spec fields, I don't think so. Hm, actually Catalog might just call out to NMS for its spec parsing needs entirely. But I think these spec changes could be larger than just adding a new field and hoping it works. I might be wrong, though. This all feels like quite a bit of scope creep to me.

> It'd be great to make it easier to update NMS to add fields to the spec, but that's probably out of scope.

Yup.

Reply:
> Catalog might just call out to NMS for its spec parsing needs entirely

Reply:
> Catalog might just call out to NMS for its spec parsing needs entirely

Reply:
> Catalog might just call out to NMS for its spec parsing needs entirely
I think that's right.

>  This all feels like quite a bit of scope creep to me.

Reply:
> Catalog might just call out to NMS for its spec parsing needs entirely

I think that's right.

> This all feels like quite a bit of scope creep to me.

We're going to have to specify the layout *somewhere*, the only question here is where. AFAICT the choices are:

* in the current spec
* in some other new spec
* in the code

Reply:
From the impromptu meeting we held 10/28 9:30am PT, we decided to go with an "internal" spec. This will be a structured document that the importer cell could read that would set the layout. This avoids having to do major updates to Narrative Method Store, and avoids hard coding layouts that might be evolving rapidly over the next few months.

Reply:
Did we make this decision? I remember talking about it but it wasn't clear to me that we actually made a decision - or rather that the engineering team made a decision.